File: app/src/main/java/rod/bailey/trafficatnsw/NavigationListAdapter.java
package rod.bailey.trafficatnsw;

import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ListAdapter;

import java.util.LinkedList;
import java.util.List;

public class NavigationListAdapter extends BaseAdapter implements ListAdapter {

	public class Item {
		private final String text;

		public Item(String text) {
			this.text = text;
		}
	}

	public class PrimaryItem extends Item {
		private final int drawableId;

		public PrimaryItem(String text, int drawableId) {
			super(text);
			this.drawableId = drawableId;
		}
	}

	public class SecondaryItem extends Item {
		public SecondaryItem(String text) {
			super(text);
		}
	}

	private final List<Item> items = new LinkedList<Item>();

	/**
	 * 
	 */
	private final MainActivity mainActivity;

	public NavigationListAdapter(MainActivity mainActivity) {
		this.mainActivity = mainActivity;

		items.add(new PrimaryItem("Incidents", R.drawable.ic_action_warning)); // [0]
		items.add(new SecondaryItem("Sydney")); // [1]
		items.add(new SecondaryItem("Regional NSW")); // [2]

		items.add(new PrimaryItem("Cameras", R.drawable.ic_action_camera)); // [3]
		items.add(new SecondaryItem("Favourites")); // [4]
		items.add(new SecondaryItem("Sydney")); // [5]
		items.add(new SecondaryItem("Regional NSW")); // [6]

		items.add(new PrimaryItem("Travel Times", R.drawable.ic_action_time)); // [7]
		items.add(new SecondaryItem("M1")); // [8]
		items.add(new SecondaryItem("M2"));// [9]
		items.add(new SecondaryItem("M4")); // [10]
		items.add(new SecondaryItem("M7")); // [11]
	}

	@Override
	public int getCount() {
		return items.size();
	}

	@Override
	public Object getItem(int position) {
		return items.get(position);
	}

	@Override
	public long getItemId(int position) {
		return position;
	}

	@Override
	public boolean isEnabled(int position) {
		Item item = items.get(position);
		return (item instanceof SecondaryItem);
	}

	@Override
	public View getView(int position, View convertView, ViewGroup parent) {
		Item item = items.get(position);

		View result = null;

		if (item instanceof PrimaryItem) {
			result = new NavigationListHeadingView(mainActivity, item.text,
					((PrimaryItem) item).drawableId, position > 0);
		} else {

			result = new NavigationListSubheadingView(mainActivity, item.text);
		}

		return result;
	}

}

File: app/src/main/java/rod/bailey/trafficatnsw/ui/view/NavigationListAdapter.kt
package rod.bailey.trafficatnsw.ui.view

import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ListAdapter
import java.util.LinkedList
import rod.bailey.trafficatnsw.MainActivity
import rod.bailey.trafficatnsw.R

class NavigationListAdapter(private val mainActivity: MainActivity) : BaseAdapter(), ListAdapter {
	open inner class Item(val text: String)

	inner class PrimaryItem(text: String, val drawableId: Int) : Item(text)

	inner class SecondaryItem(text: String) : Item(text)

	private val items = LinkedList<Item>()

	init {
		items.add(PrimaryItem("Incidents", R.drawable.ic_action_warning)) // [0]
		items.add(SecondaryItem("Sydney")) // [1]
		items.add(SecondaryItem("Regional NSW")) // [2]
		items.add(PrimaryItem("Cameras", R.drawable.ic_action_camera)) // [3]
		items.add(SecondaryItem("Favourites")) // [4]
		items.add(SecondaryItem("Sydney")) // [5]
		items.add(SecondaryItem("Regional NSW")) // [6]
		items.add(PrimaryItem("Travel Times", R.drawable.ic_action_time)) // [7]
		items.add(SecondaryItem("M1")) // [8]
		items.add(SecondaryItem("M2"))// [9]
		items.add(SecondaryItem("M4")) // [10]
		items.add(SecondaryItem("M7")) // [11]
	}

	override fun getCount(): Int {
		return items.size
	}

	override fun getItem(position: Int): Any {
		return items[position]
	}

	override fun getItemId(position: Int): Long {
		return position.toLong()
	}

	override fun isEnabled(position: Int): Boolean {
		val item = items[position]
		return item is SecondaryItem
	}

	override fun getView(position: Int, convertView: View, parent: ViewGroup): View {
		val item = items[position]
		var result: View?

		if (item is PrimaryItem) {
			result = NavigationListHeadingView(mainActivity,
																			   item.text,
																			   item.drawableId,
																			   position > 0)
		} else {
			result = NavigationListSubheadingView(mainActivity,
																				  item.text)
		}

		return result
	}
}